Refactor VBBA setup flow with new API ConnectBankAccountWithPlaid

nkuoch opened the pull request:
### Details
<!-- Explanation of the change or anything fishy that is going on -->
Use ConnectBankAccountWithPlaid instead of BankAccount_SetupWithdrawal
And some small refactoring on all steps of vbba setup
- [x] HOLD on https://github.com/Expensify/Web-Expensify/pull/34910

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/Expensify/issues/226848

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/226848

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console
- [x] Tested in https://github.com/Expensify/Web-Expensify/pull/34910

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
- [ ] Tests from https://github.com/Expensify/Web-Expensify/pull/34910

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web - Chrome
<!-- Insert screenshots of your changes on the web platform (from chrome mobile browser)-->

#### Mobile Web - Safari
<!-- Insert screenshots of your changes on the web platform (from safari mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

nkuoch committed:
Refactor VBBA setup flow with new API ConnectBankAccountWithPlaid

nkuoch committed:
Fix tests

MariaHCD commented on src/components/AddPlaidBankAccount.js:
```diff
@@ -119,30 +115,30 @@ class AddPlaidBankAccount extends React.Component {
     /**
      * Triggered when user selects a Plaid bank account.
-     * @param {String} index
+     * @param {String} plaidAccountID
```
The changes here will conflict with https://github.com/Expensify/App/pull/10953/files#diff-2a35d342f7be1cc5c1a8f02937f635a5c7083ec66b64b400027d65087d400e09 👀 

cc: @ctkochan22 

I do like the idea of using the plaidAccountID instead of the index, though.

MariaHCD commented on src/libs/actions/BankAccounts.js:
```diff
@@ -42,6 +40,62 @@ function clearPlaid() {
+    Onyx.merge(ONYXKEYS.PLAID_DATA, plaidData);
+}
+
+function getOnyxDataForVBBA() {
```
Already exists :) 

https://github.com/Expensify/App/blob/641b06cd46d420daf0861faade6dbf1a119e4487/src/libs/actions/BankAccounts.js#L106-L141

luacmartins commented on src/libs/actions/ReimbursementAccount/setupWithdrawalAccount.js:
```diff
@@ -241,12 +236,12 @@ function setupWithdrawalAccount(params) {
+            Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isLoading: false, achData: {...updatedACHData}});
             console.error(response.stack);
-            errors.showBankAccountErrorModal(Localize.translateLocal('common.genericErrorMessage'));
+            errors.showBankAccountFormValidationError('common.genericErrorMessage');
```
We still need to translate this error key, right? Otherwise we will just display `'common.genericErrorMessage'` to the user

luacmartins commented on src/pages/settings/Payments/PaymentsPage/BasePaymentsPage.js:
```diff
@@ -441,7 +441,9 @@ class BasePaymentsPage extends React.Component {
                     danger
                 />
-                <OfflineWithFeedback onClose={() => PaymentMethods.clearWalletError()} errors={this.props.userWallet.errors} errorRowStyles={[styles.ph6, styles.pv2]} />
+                <OfflineWithFeedback onClose={() => PaymentMethods.clearWalletError()} errors={this.props.userWallet.errors} errorRowStyles={[styles.ph6, styles.pv2]}>
+                    <div />
+                </OfflineWithFeedback>
```
This seems wrong, `<div>` is not supported in native platforms. We should wrap whatever component this error is related to. If this is a generic error, we should use [DotIndicatorMessage](https://github.com/Expensify/App/blob/641b06cd46d420daf0861faade6dbf1a119e4487/src/components/DotIndicatorMessage.js) instead

luacmartins commented on src/pages/ReimbursementAccount/ReimbursementAccountForm.js:
```diff
@@ -22,11 +20,22 @@ const propTypes = {
     onSubmit: PropTypes.func.isRequired,
 
+    buttonText: PropTypes.string,
+
```
NAB extra line

luacmartins commented on src/pages/ReimbursementAccount/ReimbursementAccountForm.js:
```diff
@@ -40,14 +49,9 @@ class ReimbursementAccountForm extends React.Component {
-            CONST.BANK_ACCOUNT.STEP.BANK_ACCOUNT,
-        );
+        const hasErrorFields = _.size(this.props.reimbursementAccount.errorFields) > 0;
+        const error = _.last(_.values(this.props.reimbursementAccount.errors));
```
Should we use `ErrorUtils.getLatestErrorMessage` here?

luacmartins commented on src/pages/ReimbursementAccount/ReimbursementAccountForm.js:
```diff
@@ -57,17 +61,19 @@ class ReimbursementAccountForm extends React.Component {
-                    onFixTheErrorsLinkPressed={() => {
-                        this.form.scrollTo({y: 0, animated: true});
-                    }}
-                    message={this.getErrorMessage()}
```
We can also remove `getErrorMessage` now

luacmartins commented on src/pages/ReimbursementAccount/CompanyStep.js:
```diff
@@ -331,6 +312,7 @@ CompanyStep.propTypes = propTypes;
 export default compose(
     withLocalize,
     withOnyx({
+        // Needed, to retrieve errorFields
```
```suggestion
        // Needed to retrieve errorFields
```

luacmartins commented on src/pages/ReimbursementAccount/CompanyStep.js:
```diff
@@ -159,25 +146,30 @@ class CompanyStep extends React.Component {
+            plaidAccountID: ReimbursementAccountUtils.getDefaultStateForField(this.props, 'plaidAccountID'),
+            isSavings: ReimbursementAccountUtils.getDefaultStateForField(this.props, 'isSavings'),
+            plaidAccessToken: ReimbursementAccountUtils.getDefaultStateForField(this.props, 'plaidAccessToken'),
+            acceptTerms: true,
```
NAB I don't think we should hardcode this value

MariaHCD replied:
I think we can remove this now since it's being set on the API side

luacmartins reviewed (changes requested):
Left an initial review. I still need to test this.

ctkochan22 commented on src/components/AddPlaidBankAccount.js:
```diff
@@ -187,18 +183,18 @@ class AddPlaidBankAccount extends React.Component {
                             label: this.props.translate('bankAccount.chooseAnAccount'),
                         } : {}}
-                        value={this.state.selectedIndex}
+                        value={selectedPlaidBankAacount.plaidAccountID}
```
`selectedPlaidBankAacount` is misspelled

nkuoch committed:
Fixes from PR reviews:

- Early return
- Mispelled selectedPlaidBankAccount
- Remove getOnyxDataForVBBA
- Use Localize.translateLocal for genericErrorMessage
- Use ErrorUtils.getLatestErrorMessage
- Refactor acceptTerms

nkuoch commented:
Updated!

nkuoch committed:
Clear Plaid before opening Plaid modal on personal bank account page

nkuoch committed:
Wrap OfflineWithFeedback around PaymentMethodList

luacmartins reviewed (changes requested):
While testing, I noticed that the loading spinner in the [AddPlaidBankAccount](https://github.com/Expensify/App/blob/bd1e5d2340eb6d388411dffb343bbba859d714b5/src/components/AddPlaidBankAccount.js#L137-L167) shows at the top of the form, instead of vertically centered.

![Screen Shot 2022-09-22 at 2 06 53 PM](https://user-images.githubusercontent.com/22219519/191847271-963639ac-c372-4842-a3a9-e4763606e0f6.png)

Rest LGTM.

MariaHCD commented on src/pages/ReimbursementAccount/BankAccountManualStep.js:
```diff
@@ -0,0 +1,171 @@
+    ...withLocalizePropTypes,
+};
+
+class BankAccountManualStep extends React.Component {
```
Heads up @Gonals, we are splitting the manual step into its own component. Maybe for https://github.com/Expensify/App/pull/10987, you can branch off this PR.

MariaHCD commented on src/pages/ReimbursementAccount/BankAccountManualStep.js:
```diff
@@ -0,0 +1,171 @@
+export default compose(
+    withLocalize,
+    withOnyx({
+        // Needed, to retrieve errorFields
```
```suggestion
        // Needed to retrieve errorFields
```

nkuoch committed:
Fix loader not vertically centered

MariaHCD reviewed (commented):
Just some small changes from me but otherwise, my local tests worked well 🎊

nkuoch committed:
Clean comments and remove acceptTerms

nkuoch committed:
Merge branch 'main' into nat-VBAnewAPIs

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

nkuoch commented:
^ this is a bug, all the tests were passing melvin

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3130693184) to staging by @nkuoch in version: 1.2.7-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3138719410) to production by @roryabraham in version: 1.2.7-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
